perceval/components/source.py

```diff
@@ -201,16 +181,48 @@ def generate_distribution(self, expected_input: BasicState, prob_threshold: floa
         :param prob_threshold: Probability threshold under which the resulting state is filtered out. By default,
             `global_params['min_p']` value is used.
         """
-        dist = SVDistribution()
         prob_threshold = max(prob_threshold, global_params['min_p'])
         get_logger().debug(f"Apply 'Source' noise model to {expected_input}", channel.general)
-        for photon_count in expected_input:
-            dist = SVDistribution.tensor_product(dist, self.probability_distribution(photon_count), prob_threshold)
+
+        distributions = [self.probability_distribution(photon_count, prob_threshold) for photon_count in expected_input]
+        dist = SVDistribution.list_tensor_product(distributions, prob_threshold=prob_threshold)
+
         dist.normalize()
         if self.simplify_distribution and self.partially_distinguishable:
             dist = anonymize_annotations(dist, annot_tag='_')
         return dist
 
+    def generate_samples(self, max_samples, expected_input: BasicState):
+        """
+        Simulates plugging the photonic source on certain modes and turning it on.
+        Computes the input probability distribution
+
+        :param max_samples: Number of samples to generate
+        :param expected_input: Expected input BasicState
+            The properties of the source will alter the input state. A perfect source always delivers the expected state
+            as an input. Imperfect ones won't.
+        """
+
+        bsd = self._generate_one_photon_distribution()
+
+        samples = []
+        for photon_count in expected_input:
+            if photon_count == 0 or self.is_perfect():
+                new_samples = [BasicState([photon_count])] * max_samples
+            else:
+                new_samples = bsd.sample(max_samples, non_null=False)
+                for _ in range(photon_count-1):
+                    new_samples_one_mode = bsd.sample(max_samples, non_null=False)
```

Comment:
You need to generate the bsd for each input photon since il will change the annotations. If you want, you don't need to generate it each time `if not self.partially_distinguishable`

Reply:
You still need to create a new BSD for all subsequent photons in the mode.